feat(api): Move GitHub authentication to header

slohmes opened the pull request:
Move GitHub API authentication to header, in response to GItHub deprecating API authentication via query parameters. The code updated in this PR seems to be the only place in the codebase that was authenticating via query params, but if we missed a spot, we'll be alerted by emails from GitHub and a "brownout" well before the official removal date of November 13, 2020.

More info: https://developer.github.com/changes/2020-02-10-deprecating-auth-through-query-param/

Addresses [ENT-16](https://getsentry.atlassian.net/browse/ENT-16).

Commit:
Use auth instead of access-token param

Commit:
add test

slohmes commented on tests/sentry/auth/providers/github/test_client.py:
```diff
@@ -0,0 +1,37 @@
+from sentry.auth.providers.github.client import GitHubClient
+
+
+class MockReq:
+    def __init__(self, status_code=200, content=None):
+        self.status_code = status_code
+        self.content = content
+
+
+class MockHttp:
+    def __init__(self):
+        self.calledArgs = []
+        self.calledKwargs = {}
+
+    def get(self, *args, **kwargs):
+        for arg in args:
+            self.calledArgs.append(arg)
+        self.calledKwargs = kwargs
+        return MockReq()
+
+
+@patch("sentry.utils.json.loads")
```
There wasn't much prior art I could find for auth client testing, so please lmk if you have ideas for cleaning up this test.

scefali replied:
@slohmes I would think you could just use the mock responses library we use: https://github.com/getsentry/responses

slohmes replied:
There's a library! Perfect, thank you!

EvanPurkhiser commented:
May want to check getsentry too if you haven't

slohmes commented:
@EvanPurkhiser totally forgot to check getsentry, looks like it is being used in a couple places there -- thanks for the reminder! Lemme fix them up real quick.

scefali commented on src/sentry/auth/providers/github/client.py:
```diff
@@ -21,15 +21,12 @@ def __init__(self, client_id, client_secret):
-        params = {"client_id": self.client_id, "client_secret": self.client_secret}
-
         headers = {"Authorization": "token {0}".format(access_token)}
+        auth = (self.client_id, self.client_secret)
```
@slohmes I would be surprised if we actually need to provide the client_id and client secret if we already have an `Authorization` header. Does the auth still work if you remove those parameters?

slohmes replied:
It does, thanks!

Commit:
Remove unnecessary extra auth params

Commit:
Cleanup

scefali commented on src/sentry/auth/providers/github/provider.py:
```diff
@@ -58,7 +56,7 @@ def build_identity(self, state):
     def refresh_identity(self, auth_identity):
-        client = GitHubClient(self.client_id, self.client_secret)
+        client = GitHubClient()
         access_token = auth_identity.data["access_token"]
```
@slohmes To me, it would be much more logical to make `GitHubClient` take the `access_token` as an argument since we need it for every action. Granted, it's outside the scope of the PR so I'll leave that as an optional suggestion.

slohmes replied:
That makes sense to me. Here's what that'd look like: https://github.com/getsentry/sentry/pull/19146 -- lmk what you think.

RyanSkonnord replied:
👍 to this change.

markstory commented on tests/sentry/auth/providers/github/test_client.py:
```diff
@@ -0,0 +1,32 @@
+        self.content = content
+
+
+class MockHttp:
```
You could likely use `responses` to do capture the requests made and not need to mock json.loads as well. The [docs for responses](https://github.com/getsentry/responses#dynamic-responses) have some examples, and the ones that use `responses.calls[0].request.params` are likely relevant to what you're doing.

Commit:
Cleanup test

Commit:
Whoops delete that comment

slohmes commented:
> @EvanPurkhiser totally forgot to check getsentry, looks like it is being used in a couple places there -- thanks for the reminder! Lemme fix them up real quick.

I'm going to add the getsentry changes in a separate PR, to keep these changes smaller and safer.

Other than that -- this PR should be ready for re-review. Thanks for all the feedback so far, y'all!

RyanSkonnord commented:
I think this is ready to approve. @slohmes, do you want to merge #19146 first?

Commit:
ref(api): Set access_token as GitHubClient attribute (#19146)

Instead of passing the access_token in to every method call for GitHubClient, pass it into the constructor and store it as an object attribute.
